draft-barnes-acme.md

```diff
@@ -218,304 +219,275 @@ In this section, we describe the four certificate management functions that ACME
+* "identifier" the identifier submitted by the client.
+* "key": the key used to verify the client's JWS request (i.e., the contents of the "jwk" field in the JWS header)
+* "status": SHOULD be "pending" (MAY be omitted)
+* "challenges" and "combinations": As selected by the server's policy for this identifier
+* "contact": the "contact" field submitted by the client, if provided
+* The "expires" and "recoveryToken" fields MUST be absent.
 
-type (required, string):
-: "challenge"
+The server allocates a new URI for this authorization, and returns a 201 (Created) response, with the authorization URI in a Location header field, and the JSON authorization object in the body.
 
-sessionID (required, string):
-: An opaque string that allows the server to correlate transactions related to this challenge request.
+~~~~~~~~~~
 
-nonce (required, string):
-: A base64-encoded octet string that the client is expected to sign with the private key of the key pair being authorized.
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/authz/asdf
 
-challenges (required, array):
-: A list of challenges to be fulfilled by the client in order to prove possession of the identifier.  The syntax for challenges is described in Section {{identifier-validation-challenges}}.
+{
+  "identifier": {
+    "type": "domain",
+    "value": "example.org"
+  },
 
-combinations (optional, array of arrays):
-: A collection of sets of challenges, each of which would be sufficient to prove possession of the identifier.  Clients SHOULD complete a set of challenges that that covers at least one set in this array.  Challenges are represented by their associated zero-based index in the challenges array.
+  "key": { /* JWK from JWS header */ },
 
-For example, if the server wants to have the client demonstrate both that the client controls the domain name in question, and that this client is the same client that previously requested authorization for the domain name, it might issue the following request.  The client is expected to provide "simpleHttps" and "recoveryToken" responses ("[0,2]"), or else "dns" and "recoveryToken" responses ("[1,2]"), or all three.
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
 
-~~~~~~~~~~
-{
-  "type": "challenge",
-  "sessionID": "aefoGaavieG9Wihuk2aufai3aeZ5EeW4",
-  "nonce": "czpsrF0KMH6dgajig3TGHw",
-  "challenges": [
-    {
-      "type": "simpleHttps",
+  "challenges": {
+    "simpleHttps": {
```

Comment:
This data structure change makes it difficult to issue challenges of the same type.  This isn't a problem given the current challenges and as ACME's goal is DV.  However, if there is a challenge someday where different parameters to the challenge prove different components of the authorization we will be stuck with a hack (some form of compound challenge or defining multiple challenges of the same type with different names).  This probably doesn't justify the complication involved with using a list and indices, but is something to keep in mind.